File: microservices/review/service/src/main/java/fr/edjaz/microservices/core/review/services/MessageProcessor.java
package fr.edjaz.microservices.core.review.services;

import fr.edjaz.api.core.review.Review;
import fr.edjaz.api.core.review.ReviewService;
import fr.edjaz.api.event.Event;
import fr.edjaz.util.exceptions.EventProcessingException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cloud.stream.annotation.EnableBinding;
import org.springframework.cloud.stream.annotation.StreamListener;
import org.springframework.cloud.stream.messaging.Sink;

@EnableBinding(Sink.class)
@Slf4j
public class MessageProcessor {

    private final ReviewService reviewService;

    @Autowired
    public MessageProcessor(ReviewService reviewService) {
        this.reviewService = reviewService;
    }

    @StreamListener(target = Sink.INPUT)
    public void process(Event<Integer, Review> event) {

        LOG.info("Process message created at {}...", event.getEventCreatedAt());

        switch (event.getEventType()) {

        case CREATE:
            Review review = event.getData();
            LOG.info("Create review with ID: {}/{}", review.getProductId(), review.getReviewId());
            reviewService.createReview(review);
            break;

        case DELETE:
            int productId = event.getKey();
            LOG.info("Delete reviews with ProductID: {}", productId);
            reviewService.deleteReviews(productId);
            break;

        default:
            String errorMessage = "Incorrect event type: " + event.getEventType() + ", expected a CREATE or DELETE event";
            LOG.warn(errorMessage);
            throw new EventProcessingException(errorMessage);
        }

        LOG.info("Message processing done!");
    }
}

File: microservices/review/service/src/main/kotlin/fr/edjaz/microservices/core/review/services/MessageProcessor.kt
package fr.edjaz.microservices.core.review.services

import fr.edjaz.api.core.review.Review
import fr.edjaz.api.core.review.ReviewService
import fr.edjaz.api.event.Event
import fr.edjaz.util.exceptions.EventProcessingException
import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import java.util.function.Consumer

@Configuration
class MessageProcessor {

    companion object {
        @Suppress("JAVA_CLASS_ON_COMPANION")
        @JvmStatic
        private val logger = LoggerFactory.getLogger(javaClass.enclosingClass)
    }

    @Bean
    fun sink(reviewService: ReviewService): Consumer<Event<Int, Review>> {
        return Consumer { event: Event<Int, Review> ->
            logger.info("Process message created at {}...", event.eventCreatedAt)
            when (event.eventType) {
                Event.Type.CREATE -> {
                    val review = event.data
                    logger.info("Create review with ID: {}/{}", review!!.productId, review.reviewId)
                    reviewService.createReview(review)
                }
                Event.Type.DELETE -> {
                    val productId = event.key!!
                    logger.info("Delete reviews with ProductID: {}", productId)
                    reviewService.deleteReviews(productId)
                }
                else -> {
                    val errorMessage = "Incorrect event type: " + event.eventType + ", expected a CREATE or DELETE event"
                    logger.warn(errorMessage)
                    throw EventProcessingException(errorMessage)
                }
            }
            logger.info("Message processing done!")
        }
    }
}

